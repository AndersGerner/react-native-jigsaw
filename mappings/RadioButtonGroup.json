{"name":"Radio Button Group","tag":"RadioButtonGroup","category":"button","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Control_Radio.png","props":{"options":{"label":"Options","description":"Options for the button group.","type":"arrayInput","value":[],"editable":true,"required":true},"direction":{"label":"Horizontal/Vertical","description":"Whether the buttons should be Horizontal or Vertical","type":"flatArray","value":"horizontal","options":["horizontal","vertical"],"editable":true,"required":true},"activeColor":{"label":"Active Color","description":"Color of the button when it's selected","value":null,"type":"color","editable":true,"required":true},"inactiveColor":{"label":"Inactive Color","description":"Color of the button when it's selected not selected","value":null,"type":"color","editable":true,"required":true},"contentColor":{"label":"Selected Content Color","description":"Color of the content(Icon and Label)","value":null,"type":"color","editable":true,"required":true},"unselectedContentColor":{"label":"Unselected Content Color","description":"Unfinished Color of the content(Icon and Label)","value":null,"type":"color","editable":true,"required":true},"borderColor":{"label":"Border Color","description":"Border color of the option","value":null,"type":"color","editable":true,"required":true},"labelStyle":{"label":"Label Style","description":"Font and weight of the Label","type":"typeStyle","value":"Button","editable":true,"required":true},"optionSpacing":{"label":"Option Spacing","description":"The spacing between each option","type":"number","value":1,"min":0,"max":20,"step":1,"precision":1,"editable":true,"required":false},"borderRadius":{"label":"Border Radius","description":"The border radius for the container or options","type":"number","value":0,"min":0,"max":100,"step":1,"precision":1,"editable":true,"required":false},"iconSize":{"label":"Icon Size","description":"The size of the icon if enabled","type":"number","value":0,"min":0,"max":50,"step":1,"precision":1,"editable":true,"required":false},"fieldName":{"label":"Field name","description":"The name of the field within the screen that will store this component's value","type":"fieldName","value":"radioButtonValue","valuePropName":"value","handlerPropName":"onChange","editable":true,"required":false}},"layout":{}}