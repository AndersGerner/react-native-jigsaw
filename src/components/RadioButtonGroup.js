/* @flow */

import * as React from "react"
import { withTheme } from "../core/theming"
import { View, Text } from "react-native"
import Icon from "./Icon.js"
import { COMPONENT_TYPES, FORM_TYPES } from "../core/component-types"
import Touchable from "./Touchable"

type Props = {
  style: style,
  direction: string,
  options: Array<{ id: number, icon: string, label: string }>,
  activeColor: string,
  inactiveColor: string,
  labelStyle: object,
  iconSize: number,
  contentColor: string,
  borderRadius: number,
  optionSpacing: number,
  borderColor: string,
  theme: theme
}

class RadioButtonGroup extends React.Component<Props> {
  state = { selected: this.props.defaultSelection }

  onPress = selected => {
    this.setState({ selected })
  }

  render() {
    const {
      style,
      direction,
      options,
      activeColor,
      inactiveColor,
      labelStyle,
      iconSize,
      contentColor,
      borderRadius,
      optionSpacing,
      borderColor,
      theme: { colors }
    } = this.props

    const optionWidth = style.width / options.length
    const marginHorizontal = direction === "horizontal" ? optionSpacing / 2 : 0
    const marginVertical = direction === "vertical" ? optionSpacing / 2 : 0

    return (
      <View
        style={{
          flexDirection: direction === "vertical" ? "column" : "row",
          alignItems: "center",
          borderRadius: optionSpacing ? 0 : borderRadius,
          overflow: "hidden"
        }}>
        {options.map((option, index) => {
          const selected = option.label == this.state.selected
          return (
            <Touchable onPress={() => this.onPress(option.label)}>
              <View
                style={{
                  flexDirection: "row",
                  alignItems: "center",
                  justifyContent: "center",
                  backgroundColor: selected ? activeColor : inactiveColor,
                  height: style.height,
                  width: optionWidth,
                  borderLeftWidth: borderColor && index !== 0 ? 0.5 : 0,
                  borderRightWidth: borderColor && index !== options.length - 1 ? 0.5 : 0,
                  borderColor,
                  borderRadius: optionSpacing ? borderRadius : 0,
                  marginLeft: marginHorizontal,
                  marginRight: marginHorizontal,
                  marginTop: marginVertical,
                  marginBottom: marginVertical
                }}>
                {option.icon ? (
                  <Icon name={option.icon} size={iconSize} color={contentColor} />
                ) : null}
                {option.label ? (
                  <Text style={[labelStyle, { color: contentColor }]}>{option.label}</Text>
                ) : null}
              </View>
            </Touchable>
          )
        })}
      </View>
    )
  }
}

export default withTheme(RadioButtonGroup)

export const SEED_DATA = {
  name: "Radio Button Group",
  tag: "RadioButtonGroup",
  category: COMPONENT_TYPES.formControl,
  preview_image_url: "{CLOUDINARY_URL}/Control_Radio.png",
  props: {
    options: {
      label: "Options",
      description: "Options for the button group.",
      type: FORM_TYPES.arrayInput,
      value: [],
      editable: true,
      required: true
    },
    direction: {
      label: "Horizontal/Vertical",
      description: "Whether the buttons should be Horizontal or Vertical",
      type: FORM_TYPES.flatArray,
      value: "horizontal",
      options: ["horizontal", "vertical"],
      editable: true,
      required: true
    },
    activeColor: {
      label: "Active Color",
      description: "Color of the button when it's selected",
      value: null,
      type: FORM_TYPES.color,
      editable: true,
      required: true
    },
    inactivecolor: {
      label: "Inactive Color",
      description: "Color of the button when it's selected not selected",
      value: null,
      type: FORM_TYPES.color,
      editable: true,
      required: true
    },
    contentColor: {
      label: "Content Color",
      description: "Color of the content(Icon and Label)",
      value: null,
      type: FORM_TYPES.color,
      editable: true,
      required: true
    },
    borderColor: {
      label: "Border Color",
      description: "Border color of the option",
      value: null,
      type: FORM_TYPES.color,
      editable: true,
      required: true
    },
    labelStyle: {
      label: "Label Style",
      description: "Font and weight of the Label",
      type: FORM_TYPES.typeStyle,
      value: null,
      editable: true,
      required: true
    },
    optionSpacing: {
      label: "Option Spacing",
      description: "The spacing between each option",
      type: FORM_TYPES.number,
      value: 0,
      min: 0,
      max: 20,
      step: 1,
      precision: 1,
      editable: true,
      required: false
    },
    borderRadius: {
      label: "Border Radius",
      description: "The border radius for the container or options",
      type: FORM_TYPES.number,
      value: 0,
      min: 0,
      max: 100,
      step: 1,
      precision: 1,
      editable: true,
      required: false
    },
    iconSize: {
      label: "Icon Size",
      description: "The size of the icon if enabled",
      type: FORM_TYPES.number,
      value: 0,
      min: 0,
      max: 50,
      step: 1,
      precision: 1,
      editable: true,
      required: false
    }
  },
  layout: {
    width: 200,
    height: 40
  }
}
