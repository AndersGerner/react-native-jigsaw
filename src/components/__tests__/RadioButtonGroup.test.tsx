import * as React from "react";
import renderer from "react-test-renderer";
import RadioButtonGroup from "../RadioButtonGroup";

it("renders radio button group", () => {
  const tree = renderer
    .create(
      <RadioButtonGroup
        iconSize={20}
        value="second"
        options={[{ label: "first" }, { label: "second" }]}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
