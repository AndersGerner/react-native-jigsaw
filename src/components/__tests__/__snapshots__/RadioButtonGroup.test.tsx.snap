// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders radio button group 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": undefined,
        "flexDirection": "row",
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderColor": "rgba(234, 237, 242, 1)",
            "borderLeftWidth": 0,
            "borderRadius": 0,
            "borderRightWidth": 0,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
          }
        }
      >
        <Text
          style={
            Array [
              undefined,
              Object {
                "color": undefined,
              },
            ]
          }
        >
          first
        </Text>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "opacity": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderColor": "rgba(234, 237, 242, 1)",
            "borderLeftWidth": 0,
            "borderRadius": 0,
            "borderRightWidth": 0,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
          }
        }
      >
        <Text
          style={
            Array [
              undefined,
              Object {
                "color": undefined,
              },
            ]
          }
        >
          second
        </Text>
      </View>
    </View>
  </View>
</View>
`;
